### PR TITLE
 Ancestral Value Maps : 105X

### PR DESCRIPTION
**Pull request by Sam-Harper**
This is a rebase of https://github.com/cms-sw/cmssw/pull/25576. Because I like to have the base release in the branch name (I know its easy to  find out but this makes my life a bit easier), I decided to close that PR and reopen it with a new branch. The text of that PR follows below:

Valuemaps are used extensively to associate values with objects. An object is identified via a product id/ index pair, corresponding to a given item in a collection. This is used extensively in the HLT but is also used frequently in RECO. A value map can only be used with a reference to an object from the collection it was created with.

The issue arises in that we often need to make new collections of electrons/photons (such as converting them to pat or updating the objects with new embedded information). Currently when you do this, you need to rekey all the value maps associated with that collection, which to put it mildly is a pain.

This update allows electrons and photons to track which objects they were made from by storing a edm::Ptr to the original object. edm::ValueMap interface has been extended to recognize when this information is available and if the edm::Ptr is invalid, cycle through its parents to see if it finds one that matches the collection it was created from. This means a new collection of pat::Electrons can use value maps keyed to the orginal GsfElectrons. This greatly simplifies the code, look what happens to RecoEgamma/EgammaTools/interface/EGExtraInfoModifierFromValueMaps.h. ReducedEGProducer will similarly get smaller as it spends some effort rekeying valuemaps. Also makes VID so much easier.

The setting of the parents references is optional and up to the code which creates the new electrons/photons. You may not want to do it for temporary collections for which no value maps are produced.

Note the code refers to parentRefs etc but perhaps I should rename it to pointers as thats what they are.

Finally, this is probably of interest to other POGs who likely have similar issues. As such I'm open to the idea allowing reco::Candidates to track their parents and adding this capability for everybody.

Btw backporting some of this to 102X would be useful, this would make E/gamma user recipes easier. However I'm not going to push for this.

Also while I was compiling half of CMSSW anyways, I added an "unInitialized" method to InputTag, can remove if you like but it made the code a bit more clear.

### CODE REVIEW

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by Dr15Jones** [comment]
@wddgit do you think this has any relation to slimming?

File: FWCore/Utilities/interface/InputTag.h [hunk 1]
```diff
@@ -39,6 +39,8 @@ namespace edm {
     ///product with the label and instance
     std::string const& process() const {return process_;} 
 
+    bool isUnitialized() const {return label_.empty() && instance_.empty() && process_.empty();}
```
Comment (Dr15Jones) [inline FWCore/Utilities/interface/InputTag.h]: In practice, you only have to test `label_.empty()` since a valid InputTag must always have at least that one set.

**Review by Dr15Jones — CHANGES_REQUESTED** [review]

File: FWCore/Utilities/interface/InputTag.h (hunk 1 above)
Comment (Sam-Harper) [reply FWCore/Utilities/interface/InputTag.h]: thats a good point, should have realised it, will make the change in the next update. This makes a little less of an important change as I could just do label().empty(). I'm totally neutral on this change (hence why its a seperate commit), I thought it was nice to have a little syntax sugar to telegraph the intention but I dont mind if others feel label().empty() is the way to go and avoid polluting the number of methods.

**Comment by makortel** [comment]
We had a lengthy discussion with @Dr15Jones and @wddgit, and our suggestion would be that instead of the full "parentage" of copies, the `GsfElectron`/`Photon` would store the `edm::ProductID` of the "first" product, and the index within the container. That ID would then be used to fill and read from the `ValueMap`.

More concretely
* we would create a new class `edm::UniqueElementID` that holds `edm::ProductID` and an `unsigned int` for the index
* `GsfElectron`/`Photon` would have an `edm::UniqueElementID` member
   * `edm::UniqueElement` would have size of 8 bytes, compared to 24 bytes of an empty `std::vector`, and would be much cheaper than a non-empty `std::vector<edm::Ptr<T>>`
* we would add a helper function to set these unique element IDs (for all elements of the `vector`) and put the product to `edm::Event`
* `ValueMap` would check (similar to this PR) whether the `RefKey` has (e.g.) `uniqueElementID()` method defined
  * if yes, use that as the key
  * if not, use the `RefKey` as the key

@Sam-Harper What do you think, would the suggestion above fulfill your need? If yes, the framework group would take care of adding the `edm::UniqueElementID` and the helper function.

File: RecoEgamma/EgammaTools/interface/EGExtraInfoModifierFromValueMaps.h [hunk 2]
```diff
@@ -51,72 +40,77 @@ namespace egmodifier{
...
-  ~EGXtraModFromVMObjFiller()=delete;
+//our little helper classes
+namespace egmodifier {
+  template<typename MapType>
+  class ValueMapData {
+  public:
+    ValueMapData(std::string name,const edm::InputTag& tag):name_(std::move(name)),tag_(tag){}
+    
+    void setHandle(const edm::Event& evt){if(!token_.isUninitialized()) evt.getByToken(token_,handle_);}
+    void setToken(edm::ConsumesCollector& cc){if(!tag_.isUnitialized()) token_ = cc.consumes<edm::ValueMap<MapType> >(tag_);}
```
[the remark this answers is not in the document]
Comment (Sam-Harper) [reply RecoEgamma/EgammaTools/interface/EGExtraInfoModifierFromValueMaps.h]: hi @makortel, I'm not sure there is a fundamental reason why this is not done. I tried to think of one but couldnt. But this is the interface we have unfortunately.

**Comment by Sam-Harper** [comment]
Hi @makortel,  @Dr15Jones and @wddgit

3 line summary at the bottom

Thank you all for discussing this, I think your idea is much better as you are fixing the problem at its core rather than my workaround. Although of course one has to see the implementation details. 

As you realised, what we want is a way to uniquely identify an electron or other object and I was sort of abusing the edm::Ptrs to do that. You're right, theres no need to store every edm::Ptr to every copy of the electron (there may be some fringe benefits in niche cases but I'm pretty sure I'll never want to access the parent electron via this pointer) and this telegraphs the intention better. 

I can see how this would be very easy to set in an automatic way as well. 

So when creating the value map, it always keys to the original collection now and for objects which dont support this, its transparent. 

First feature  I need. Am I allows to still keep the ability to pass an edm::Ref to the operator[] and have this work? This is a critical feature, while would re-writing every call to an electron / photon valuemap in CMSSW is something to be avoided, the bigger concern is that this needs to work for all private code too and here it would be good to keep changes to a minimum (ideally none). 

Second feature I need. The main reason I'm doing this is that post reco, e/gamma often does a lot of calibrations and ID embedding and the simplification of the reco code is more a side benefit. Because of the various remaking of collections, value maps are an absolute nightmare to deal with. So I need this to work on objects which dont originally have this set, ie 102X, 94X recoed objects and valuemaps.  So as long as we're fine with "original object" for these releases being the object in saved in the AOD/MINIAOD, I'm happy.

The nice advantage of this approach is that I dont need to ever generate references to use valuemaps, I can just use the operator[] with edm::UniqueElementID in future code. 

Finally one downside, doing this way removes the ability to query if this valuemap was actually produced using the current electron collection. The main use of this ability would be to catch user errors (say a user really wants to make sure they are using the correct value map, perhaps in a debuging scenario) but I think the use of this is likely to be insignificant and we leave it to the users to ensure they have typed the right input label.  But just throwing that out there. 

Let me think about this a little longer but I think your solution is best. I'm also pinging other pogs if they want this capability on their objects, so it might be done at the reco::RecoCandidate level. 

ps it occurs to me that you could also use this new class to generate an edm::Ptr / edm::Ref to the object from itself if you added a separate member to store it (ie one to the orginal object, one to the current object).  This could be useful, particularly when you have a reco::GsfElectron& in your function but you really need a reco::GsfElectronRef. 

tl;dr: I really like your idea. I want to be sure I keep the passing in of a Ref to a valuemap though. I would like a little more time to think it through and also check with other pogs. I would also like this to work on existing 102X, 94X data.

File: RecoEgamma/EgammaTools/interface/EGExtraInfoModifierFromValueMaps.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (makortel) [reply RecoEgamma/EgammaTools/interface/EGExtraInfoModifierFromValueMaps.h]: Maybe we should fix the interface then (but not in this PR).

**Comment by makortel** [comment]
@Sam-Harper 

> So when creating the value map, it always keys to the original collection now and for objects which dont support this, its transparent.

Correct. In principle one can also create a value map from the copied objects, but using the original collection as the keys.

> Am I allows to still keep the ability to pass an edm::Ref to the operator[] and have this work?

Passing `edm::Ref`/`Ptr` will (continue to) use the ref's `ProductID`+index as the key, so to use the original ID one has to explicitly call `valueMap[ref->originalElementID()]`. There are several reasons why we believe this is better
* The current behavior of `ValueMap` is preserved. We are concerned that making `ValueMap` aware of `originalElementID()` would change current behavior in unexpected ways in certain cases, e.g.
   * Currently `ValueMap` interface depends only on properties on `Ref<T>`/`Ptr<T>`. We believe it is better to avoid adding dependence also to the interface of `T`. E.g. what if some object would have two element IDs?
   * What if the `Ref` is not referenceable? (e.g. pointed-to-object was not kept)
   * What if `Ref` or `T` would not be a valid key, and calling `originalElementID()` would lead to unnecessary read from disk (or over network)?
* If we would make `ValueMap` aware of this feature, the behavior of `ValueMap` would become dependent on the `edm::Ptr` type: if accessed with `Ptr<GsfElectron>`, get a value, if accessed with `Ptr<Candidate>` (pointing to a `GsfElectron`), get an exception of incorrect key.
   * The `Ptr<Candidate>` giving "wrong" behavior happens with this PR as well
* Interface consistency wrt. `Association*` classes
* With `uniqueElementID()` clients of `ValueMap` can (likely) simplify their code by avoiding the use of `Ref`/`Ptr` altogether

> This is a critical feature, while would re-writing every call to an electron / photon valuemap in CMSSW is something to be avoided, the bigger concern is that this needs to work for all private code too and here it would be good to keep changes to a minimum (ideally none).

Unfortunately that would be the price to pay. On the other hand, now the users would have to become aware of how the way to "uniquely identify an electron" changes.

We would assist in changing all relevant code in CMSSW.

> So I need this to work on objects which dont originally have this set, ie 102X, 94X recoed objects and valuemaps. So as long as we're fine with "original object" for these releases being the object in saved in the AOD/MINIAOD, I'm happy.

Dealing with older objects would work, if the copying modules check the `originalElementID()` of the copied object, and if it is unset, set it to the value of the `Ref`/`Ptr`-to-copied-from object.

But, unfortunately, I'm not sure about the feasibility to backport to 102X or 94X as the change requires a change in data formats. I think it would be good to get a decision on this point rather sooner than later. @Dr15Jones @slava77 @perrotta @fabiocos 

> Finally one downside, doing this way removes the ability to query if this valuemap was actually produced using the current electron collection.

By keeping `ValueMap` unaware of the `originalElementID()` this ability is actually preserved.

> ps it occurs to me that you could also use this new class to generate an edm::Ptr / edm::Ref to the object from itself if you added a separate member to store it (ie one to the orginal object, one to the current object). 

This is indeed something that would become possible. On the other hand, we should not include that in the interface of `GsfElectron` etc, because then it would become too easy to create a `Ptr` with key pointing to the original object while pointing to a copied object.

**Comment by Sam-Harper** [comment]
Okay so I think we need to discuss this further then. I understand where you are coming from from a framework point of view. However what you propose puts a significant user burden which you will also have to support and we need to decide how to balance the two things. 

One small thing, you didnt correctly understand my comment about 102X , 94X. While I would love, absolutely love a backport, I do understand that there may be good reasons not to do so. However the reco group has enacted a policy that analysis is only centrally supported "out of the box" (as in a clean release with no patches)  in the latest release physics release. Therefore 105X must be able to read 102X, 94X. 

Unfortunately, I am now traveling back to my institute for our annual staff meeting so wont be able to discuss this till next week when I return.

File: RecoEgamma/EgammaTools/interface/EGExtraInfoModifierFromValueMaps.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (makortel) [reply RecoEgamma/EgammaTools/interface/EGExtraInfoModifierFromValueMaps.h]: I went and moved passing the `ConsumesCollector` to the constructor in #25617.

**Comment by makortel** [comment]
We felt that a class for "`ProductID` + index" would be useful anyway, so I added one (`edm::ElementID`) here #25618.

**Comment by slava77** [comment]
@Sam-Harper @makortel 
please clarify on the plan for the next steps with this feature/PR.
Thank you.

**Comment by Sam-Harper** [comment]
Well I think its clear it wont make 10_5_0.  We need to have a discussion on this because we are currently far appart on this and we have more pressing issues. Also we simply dont have time to re-write every valuemap call in CMSSW before 10_5_0 if we go for @makortel's solution. So I plan to focus on getting things that will make 10_5_0 before coming back to this.

**Comment by slava77** [comment]
@Sam-Harper 
perhaps this can even be closed and reopened when an appropriate solution is available.

**Comment by Sam-Harper** [comment]
okay we'll do that then
